2020 updates to Lab 1 (#24)
uchicago-cs/cmsc22000: website/content/labs/lab1.md

{{% /warning %}}

In some of your previous CS courses, you may have become familiar with using Subversion (SVN) to handle revision control (or *version control*) of your projects and assignments, and to submit your work for grading. In this class, we will be using Git, a more widespread version control system that is more commonly used in a variety of software projects.

Like SVN, Git is a *version control system* that maintains files in a *repository* that contains not just files, but also a record of all the changes made to those files. Git tracks every version of a file or directory using *commits*. When you have made changes to one or more files, you can logically group those changes into a "commit" that gets added to your repository. You can think of commits as “checkpoints” in your work, representing the work you’ve done since the previous checkpoint. This mechanism makes it possible to look at and even revert to older versions of a file by going back to your code as it was when you “checkpointed” it with a commit.

> For each lab assignment, a Git repository will be created for you on [GitHub](https://github.com/), a web-based hosting service for Git repositories. However, before that repository can be created for you, you need to have a GitHub account. If you do not yet have one, you can get an account here: https://github.com/join. Once you create your account, you may want to get the [Student Developer Pack](https://education.github.com/pack), which will give you access to a lot of other features (please note that having the Student Developer Pack is not necessary for CS 220; it's just a nice benefit you get as a student) 

Reviewer: huge mood that we're no longer going to use both GL and GH

> For each lab assignment, a Git repository will be created for you on [GitHub](https://github.com/), a web-based hosting service for Git repositories. However, before that repository can be created for you, you need to have a GitHub account. If you do not yet have one, you can get an account here: https://github.com/join. Once you create your account, you may want to get the [Student Developer Pack](https://education.github.com/pack), which will give you access to a lot of other features (please note that having the Student Developer Pack is not necessary for CS 220; it's just a nice benefit you get as a student) 

Reviewer: ```suggestion
For each lab assignment, a Git repository will be created for you on [GitHub](https://github.com/), a web-based hosting service for Git repositories. However, before that repository can be created for you, you need to have a GitHub account. If you do not yet have one, you can get an account here: https://github.com/join. Once you create your account, you may want to get the [Student Developer Pack](https://education.github.com/pack), which will give you access to a lot of other features (please note that having the Student Developer Pack is not necessary for CS220; it's just a nice benefit you get as a student) 
```

Reviewer: idk why it's not highlighting the suggestion - all I did was remove the space between `CS` and `220` (i.e. `CS 220` -> `CS220`)

Author: That is inconsistent with how we write the course name every where else, including on the front page of the website:

![image](https://user-images.githubusercontent.com/448918/78728253-7d72f900-78fc-11ea-86d5-cce63236fa48.png)

## Where should you do this lab?

For this lab, we will specifically need you to work on a CS machine, which means you will have to connect remotely to CS environment. You can do this in one of two ways:

- Using SSH, which will allow you to open a command-line terminal on a CS machine. The CS techstaff provides [detailed instructions](https://howto.cs.uchicago.edu/remote_access) on how to do this.
- Connecting to a [Virtual Desktop](https://howto.cs.uchicago.edu/techstaff:vdesk). This will show you, on your computer, effectively the exact same desktop you would see if you physically logged into a CSIL machine. However, this option may not work well for low-bandwidth Internet connections. 

> In future labs, there will be more flexibility on where you can do your work, including working locally on your own computer.

Reviewer: is this virtualbox? are the instructions I wrote up on the techstaff wiki still good?

Reviewer 2: Not virtualbox- techstaff introduced a website this quarter that allows you to connect to a graphical interface through a web browser (it's actually pretty cool, but I still prefer ssh/vbox).

The instructions for setting up vm's on the techstaff wiki are still great; I'm just assuming Borja didn't wanna recommend setting up the VM and then throwing the students down a rabbit hole. However, I was wondering why the Vm wasn't recommended here?

Reviewer 3: Queue in "my virtual machine doesn't work" messages on Slack

Author: We're going to point them to the VM in the developer's guide. For this first lab, the lack of VM mentions is intentional; I think it's important they do at least one exercise in a CS environment so they're clear on how to access it remotely (there will likely be a fair number of students who have never used SSH because they were never placed in a position to do so)

  cd cs220
  ```

* Inside that folder, create a file called `README.md` with your full name.

* On your repository's GitHub page, right under "Quick setup — if you’ve done this kind of thing before" there is a URL field with two buttons: HTTPS and SSH. Make sure to change this to "HTTPS".

> Now, from inside your CMSC 22000 directory, run the commands that appear under "...or create a new repository on the command line" *except* the first one (the one that starts with `echo`. If any commands asks you to enter a username and password, just enter your GitHub username and password. Don't worry about what each individual command does; we will be seeing what most of these commands do in this lab. However, if you encounter any error messages, please make sure to let us know.

Author: I've called this out more explicitly (and also pointed out that this is distinct from your CNetID/password)

  cd cs220
  ```

> * Inside that folder, create a file called `README.md` with your full name.

Reviewer 3: Instead of `with your full name`, maybe we should change the wording to something along the lines of `and add your full name to the file`. Really minor, but the original wording seemed a bit vague to me when I read it for the first time.

Author: Pushed a fix for this.

* On your repository's GitHub page, right under "Quick setup — if you’ve done this kind of thing before" there is a URL field with two buttons: HTTPS and SSH. Make sure to change this to "HTTPS".

> Now, from inside your CMSC 22000 directory, run the commands that appear under "...or create a new repository on the command line" *except* the first one (the one that starts with `echo`. 

Reviewer: ```suggestion
Now, from inside your CMSC 22000 directory, run the commands that appear under "...or create a new repository on the command line" *except* the first one (the one that starts with `echo`). 
```

> *Note: If any commands asks you to enter a username and password, just enter your GitHub username and password* (do not enter your CNetID and password)

Reviewer: might be worth noting that if they have 2FA enabled on their GH account, they'll need to set up a CLI token to be able to use HTTPS authentication. [here's](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line) GH's instructions - for step 7 they'll just need the `repo` permission as suggested. might also be worth throwing in a mention to set `credential.helper` to `store` so they don't have to keep the token anywhere.

Reviewer: (also - I personally use SSH auth with GH because I got sick of this flow. not sure if that's the right move for the class though 🤷‍♂)

Reviewer 2: ssh is the bomb, screw https

Author: I really don't want to unleash SSH keys on them this early. Task 8 actually asks them to figure out how to avoid having to enter their password every time (which is in line with having this lab also being about them figuring stuff out on their own)

 
Don't worry about what each individual command does; we will be seeing what most of these commands do in this lab. However, if you encounter any error messages, please make sure to let us know.

You can verify that your repository was correctly set up by going back to your repository's page on GitHub, you should now see it contains a `README.md` file. If you click on it, you can see its contents.

## Creating a commit

If you make changes to your repository, the way to store those changes (and the updated versions of the modified files) is by creating a *commit*. So, let's start by making some changes:

* Edit `README.md` to also include your CNetID
* Add a new file called `test.txt`. This file should contain a single line with the text `Hello, world!`

> Creating a commit is a two-step process. First, you have to indicate what files you want to include in your commit. Let’s say we want to create a commit that only includes the updated `test.txt` file. We can specify this operation explicitly using the `git add` command from the terminal:

Reviewer 3: Here the write-up talk about only including the `test.txt` file, but then continues to add `README.md`. As far as I understand, this sentence should say `README.md` instead of `test.txt` (or vice versa; the continuing section of the write-up should be changed to add and commit `test.txt`)

Author: Pushed a fix for this.

## Working from multiple locations

So far, you have a local repository in your CS home directory, which you have been uploading to GitHub using the `git push` command. However, if you work from multiple locations (e.g., on a CS machine but also from your laptop), you will need to be able to create a local repository in those locations too. You can do this by running the `git clone` command (don't run this command just yet):

    git clone https://github.com/cmsc22000-labs/2020-lab1-username.git

This will create a local repository that "clones" the version of the repository that is currently stored on GitHub. Let's do this in a different directory on the machine you are on. For example:

    mkdir -p /tmp/$USER/cs220
    cd /tmp/$USER/cs220
    git clone https://github.com/cmsc22000-labs/2020-lab1-username.git

Make sure to replace `username` with your GitHub username!

Take into account that, when you run `git clone`, the repository is not cloned *into* the current directory. Instead, a *new* directory (with the same name as the repository) will be created in the current directory, and you will need to `cd` into it to use Git commands for that repository.

> Now, in the local repository in your home directory, add a line to `test.txt` with the text `One more change!`. Create a commit for that change and push it to GitHub (you should know how to do this by now, but make sure to ask for help if you're unsure of how to proceed).

Reviewer 3: As these labs used to be in-person, the part where it says "make sure to ask for help" made more sense, but as we've moved onto only using Zoom and Slack, I think we should expand upon where they should ask for help if they are stuck with the lab.

Now that I think of it, I'm not even sure where they should ask for help about an issue like this. I assume #general on Slack? Or DMing one of us?

Reviewer: I feel like a dedicated slack channel with rigid threading etiquette would be the right move here

Author: As far as Slack support is concerned, there will be a #labs channel for questions about the labs (where questions will have to be threaded). We're also going to have office hours over Zoom, but we're still trying to figure out how those will work exactly. Once we figure all this out, I'll be adding a "Getting Help" section to the syllabus (but will also explain how the #labs channel will work when I release the lab)

> Next, in the *second* local repository (the one you just created either on a separate location in the machine you're working on), check if that change appears in the `test.txt` file. It will not, because you have not yet downloaded the latest commits from the repository. You can do this by running this:

Reviewer: ```suggestion
Next, in the *second* local repository (the one you just created in a separate location in the machine you're working on), check if that change appears in the `test.txt` file. It will not, because you have not yet downloaded the latest commits from the repository. You can do this by running this:
```

Before this lab's deadline, you must also complete the tasks below. Some of them can be done just with what you have learned in this lab, but most of them will require that you find the exact Git command (or series of Git commands) on your own. This is a very useful skill to develop: most software developers never take a course on Git or read a full book on Git before starting to use it; they learn the basics (like you did in this lab), and then rely on online documentation to fill the gaps.

So, for the following tasks, you are allowed to obtain the answers in any way you want **EXCEPT** by asking someone (other than a CS 220 instructor or TA) to help you. This means you cannot ask for hints, solutions, pointers to documentation, etc. from *anyone* (classmates, roommates, friends, parents, etc.). Please note that you are welcome to take the answer verbatim from a website, online reference, online forum, etc. as long as you provide *attribution* (i.e., you need to tell us where you found the answer). Of course, you must also follow the instructions you find in those references to complete the task you've been given.

> 

Reviewer 2: Couldn't comment directly on it for some reason

### Submitting your lab

Before submitting, make sure you've committed and pushed your changes to the `tasks.txt` file (remember you can run `git status` to check this). 

> TODO: Gradescope instructions

Reviewer: no TODOs on `master`! 😡

Author: That's why this PR hasn't been merged yet :-P
